### PR TITLE
FP S3603: A function that returns a Task, is not a pure function but an async void

### PR DESCRIPTION
**Pull request by Corniel**
Extending the logic was straightforward.

### CODE REVIEW

**Comment by Corniel** [comment]
@costin-zaharia-sonarsource I assume, the documentation should also be extended, mentioning this case.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PureAttributeOnVoidMethod.cs [hunk 1]
```diff
@@ -21,6 +21,17 @@ void ConfigureAge(int age)
             this.age = age;
         }
 
+        [Pure] // Noncompliant
+        Task TaskDoesNotReturn(int input)
+        {
+            return Task.FromResult(input);
+        }
+		[Pure] // Noncompliant
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PureAttributeOnVoidMethod.cs:29]: pedantic: replace tabs with spaces (also below and in the vb file)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PureAttributeOnVoidMethod.cs [hunk 2]
```diff
@@ -21,6 +21,17 @@ void ConfigureAge(int age)
             this.age = age;
         }
 
+        [Pure] // Noncompliant
+        Task TaskDoesNotReturn(int input)
+        {
+            return Task.FromResult(input);
+        }
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PureAttributeOnVoidMethod.cs:28]: pedantic: please insert an empty line between the methods

File: analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs [hunk 3]
```diff
@@ -52,10 +52,16 @@ public abstract class PureAttributeOnVoidMethodBase<TSyntaxKind> : SonarDiagnost
...
-            symbol is IMethodSymbol { ReturnsVoid: true } methodSymbol
-                   && !methodSymbol.Parameters.Any(p => p.RefKind != RefKind.None)
+            symbol is IMethodSymbol methodSymbol
+                   && (ReturnsVoid(methodSymbol) || ReturnsTask(methodSymbol))
                    && methodSymbol.GetAttributes().FirstOrDefault(a => a.AttributeClass.Is(KnownType.System_Diagnostics_Contracts_PureAttribute)) is { } pureAttribute
                 ? pureAttribute
                 : null;
+
+        private static bool ReturnsVoid(IMethodSymbol methodSymbol) =>
+            methodSymbol.ReturnsVoid && !methodSymbol.Parameters.Any(p => p.RefKind != RefKind.None);
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs:62]: As far as I can tell, the RefKind comparison should be done also for methods returning tasks.

Can you please add a test case with:
```
[Pure]
Task TaskDoesNotReturn(int input, out int ret)
{
    ret = input;
    return Task.FromResult(input);
}
```

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs (hunk 3 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs]: You're right. I added both the tests as the implementation for covering it.

**Review by costin-zaharia-sonarsource — COMMENTED** [review]
Overall it looks good to me. I've left some minor comments and the coverage is a bit low (one branch is not covered). 

It would be great if you could take a look.

File: analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs [hunk 4]
```diff
@@ -52,10 +52,18 @@ public abstract class PureAttributeOnVoidMethodBase<TSyntaxKind> : SonarDiagnost
...
-                ? pureAttribute
-                : null;
+            symbol is IMethodSymbol method
+                && NoOutParameters(method)
+                && (method.ReturnsVoid || ReturnsTask(method))
+                && GetPureAttribute(method) is { } pureAttribute
+            ? pureAttribute
+            : null;
+
+        private static bool NoOutParameters(IMethodSymbol method) => method.Parameters.All(p => p.RefKind == RefKind.None);
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs:62]: I think it should be here.

File: analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs [hunk 5]
```diff
@@ -52,10 +52,18 @@ public abstract class PureAttributeOnVoidMethodBase<TSyntaxKind> : SonarDiagnost
...
+
+        private static bool ReturnsTask(IMethodSymbol method) => method.ReturnType.Is(KnownType.System_Threading_Tasks_Task);
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs:64]: If you think they are too long for a single line, I can move the body. I thought they where short enough.

[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs:64]: This is an internal convention, if the expression has a dot, it goes to the next line 😄

File: analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs:62]: That being said, there are ony 3 values in this enum: None, Out, ByRef, at least, that is what is shown when I navigate to its definition.

[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs]: The problem is the compiler version we reference. Some of the entries were later introduced: https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.refkind?view=roslyn-dotnet-3.11.0.

What we did in the past for similar cases was to create a new enum with the same name and "Ex" suffix and add the missing member with the hardcoded value, 3 in this case (which is the same in this case for both In and RefReadOnly - I've actually checked the implementation suspecting that the doc is wrong. It wasn't 😄 ).

It's not mandatory though to handle this case. Up to you.

**Review by costin-zaharia-sonarsource — APPROVED** [review]

File: analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs]: Added `In`.

**Comment by costin-zaharia-sonarsource** [comment]
🎉 Thanks @Corniel for your contribution!

**Comment by Corniel** [comment]
@costin-zaharia-sonarsource as always, you're welcome.

**Pull request merged**